File: app/src/main/java/com/jakebarnby/filemanager/sources/local/LocalSource.java
package com.jakebarnby.filemanager.sources.local;

import android.Manifest;
import android.content.Context;
import android.os.AsyncTask;
import android.os.Environment;

import com.jakebarnby.filemanager.R;
import com.jakebarnby.filemanager.sources.models.Source;
import com.jakebarnby.filemanager.sources.SourceListener;
import com.jakebarnby.filemanager.sources.models.SourceFile;
import com.jakebarnby.filemanager.sources.models.SourceType;
import com.jakebarnby.filemanager.util.Constants;
import com.jakebarnby.filemanager.util.TreeNode;
import com.jakebarnby.filemanager.util.Utils;

import java.io.File;

/**
 * Created by jakebarnby on 2/08/17.
 */

public class LocalSource extends Source {

    private String rootPath;

    public LocalSource(String sourceName, String rootPath, SourceListener listener) {
        super(SourceType.LOCAL, sourceName, listener);
        this.rootPath = rootPath;
    }

    @Override
    public void authenticateSource(Context context) {
        mSourceListener.onCheckPermissions(Manifest.permission.WRITE_EXTERNAL_STORAGE, Constants.RequestCodes.STORAGE_PERMISSIONS);
    }

    @Override
    public void loadSource(Context context) {
        if (!isFilesLoaded()) {
            new LocalLoaderTask(this, mSourceListener)
                    .executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, rootPath);
        }
    }

    @Override
    public void logout(Context context) {

    }
}

File: app/src/main/java/com/jakebarnby/filemanager/sources/local/LocalSource.kt
package com.jakebarnby.filemanager.sources.local

import android.Manifest
import android.content.Context
import com.jakebarnby.filemanager.models.Source
import com.jakebarnby.filemanager.models.SourceConnectionType
import com.jakebarnby.filemanager.models.SourceType
import com.jakebarnby.filemanager.ui.sources.SourceFragmentContract
import com.jakebarnby.filemanager.util.Constants.RequestCodes
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.asExecutor

/**
 * Created by jakebarnby on 2/08/17.
 */
class LocalSource(
    private val rootPath: String?,
    private val presenter: SourceFragmentContract.Presenter
) : Source(
    SourceConnectionType.LOCAL,
    SourceType.LOCAL.id,
    presenter.prefsManager
) {

    override fun authenticate(context: Context) {
        presenter.onCheckPermissions(Manifest.permission.WRITE_EXTERNAL_STORAGE, RequestCodes.STORAGE_PERMISSIONS)
    }

    override fun loadFiles(context: Context) {
        if (isFilesLoaded) {
            return
        }
        LocalLoaderTask(this, presenter)
            .executeOnExecutor(Dispatchers.IO.asExecutor(), rootPath)
    }

    override fun logout(context: Context) {}

}